File: wimf-processor/src/Post/Core/UseCases/SearchPost/SearchPostRequest.cs
using Utils;

namespace PostApi
{
    public class SearchPostRequest
    {
        public const int FromMinValue = 0;
        public const int SizeMaxValue = 500;

        public static SearchPostRequest Create(
            PostFilters filters,
            int from = FromMinValue,
            int size = SizeMaxValue
        )
        {
            if (from < FromMinValue)
            {
                throw new PostException(PostExceptionType.SearchPostRequestInvalid);
            }

            if (size > SizeMaxValue)
            {
                throw new PostException(PostExceptionType.SearchPostRequestInvalid);
            }
            
            return new SearchPostRequest(filters, from, size);
        }

        public readonly int From;

        public readonly int Size;

        public readonly PostFilters Filters;

        private SearchPostRequest(
            PostFilters filters,
            int from,
            int size
        )
        {
            From = from;

            Size = size;

            Filters = filters;
        }
    }
}

File: wimf-site/src/core/domain/Post/SearchPostRequest.ts
import queryString from 'query-string';
import {PostFilters} from "./PostFilters";

interface Payload {
    readonly from?: number;
    readonly size?: number;
    readonly postFilters?: PostFilters;
}

export class SearchPostRequest {
    static defaultFrom = 0;
    static defaultSize = 10;

    static create(request?: Payload) {
        return new SearchPostRequest(request || {});
    }

    static queryString(search: SearchPostRequest) {
        return queryString.stringify(
            {
                from: search.from,
                size: search.size,
                ...search.filters,
            },
            { skipNull: true },
        );
    }

    readonly from: number;
    readonly size: number;
    readonly filters?: PostFilters

    private constructor({ from, size, postFilters }: Payload) {
        this.from = from || SearchPostRequest.defaultFrom;
        this.size = size || SearchPostRequest.defaultSize;

        if (postFilters) {
            this.filters = postFilters;
        }
    }
}
